File: LeetCode/滑动窗口/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    unordered_set<char> mp;
    int lengthOfLongestSubstring(string s) {
        // mp.clear();
        if (s.size() == 0)
            return 0;
        int l = 0,len = 0;
        for(int i = 0;i < s.size(); i++){
            char c = s[i];
           
            while(mp.find(c) != mp.end()){
                mp.erase(s[l]);
                l++;
            }
            mp.insert(c);
            len = max(len, i - l + 1);         
        }
        return len;
    }
};
// @lc code=end

int main()
{
    Solution s;
    // cout << s.lengthOfLongestSubstring("abcabcbb") << endl;
    // cout << s.lengthOfLongestSubstring("bbbbb") << endl;
    cout << s.lengthOfLongestSubstring("pwwkew") << endl;
    cout<<s.lengthOfLongestSubstring("au")<<endl;
    return 0;
}

File: LeetCode/滑动窗口/3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#

# @lc code=start
from collections import Counter


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # 时间复杂度 O(N)
        # 空间复杂度 O(len(set(s)))
        ans = 0
        cnt = Counter()
        left = 0
        for right, c in enumerate(s):
            cnt[c] += 1
            while cnt[c] > 1:
                cnt[s[left]] -= 1
                left += 1
            ans = max(ans, right-left+1)
        return ans        
# @lc code=end

s = "abcabcbb"
print(Solution().lengthOfLongestSubstring(s))